turbopack/crates/turbopack-ecmascript/src/global_module_id_strategy.rs

```diff
@@ -164,58 +164,38 @@ pub async fn children_modules_idents(
     Ok(PreprocessedChildrenIdents { modules_idents }.cell())
 }
 
+const JS_MAX_SAFE_INTEGER: u64 = (1u64 << 53) - 1;
+
 // Note(LichuAcu): This could be split into two functions: one that merges the preprocessed module
 // ids and another that generates the final, optimized module ids. Thoughts?
 pub async fn merge_preprocessed_module_ids(
     preprocessed_module_ids: Vec<Vc<PreprocessedChildrenIdents>>,
 ) -> Result<HashMap<RcStr, Vc<ModuleId>>> {
-    let mut module_id_map: HashMap<RcStr, Vc<ModuleId>> = HashMap::new();
-    let mut used_ids: HashSet<u64> = HashSet::new();
+    let mut merged_module_ids = HashMap::new();
 
     for preprocessed_module_ids in preprocessed_module_ids {
         for (module_ident, full_hash) in preprocessed_module_ids.await?.modules_idents.iter() {
-            process_module(
-                module_ident.clone(),
-                *full_hash,
-                &mut module_id_map,
-                &mut used_ids,
-            )
-            .await?;
+            merged_module_ids.insert(module_ident.clone(), *full_hash);
         }
     }
 
-    Ok(module_id_map)
-}
+    // 5% fill rate, as done in Webpack
+    // https://github.com/webpack/webpack/blob/27cf3e59f5f289dfc4d76b7a1df2edbc4e651589/lib/ids/IdHelpers.js#L366-L405
+    let optimal_range: u64 =
+        std::cmp::min((merged_module_ids.len() * 20) as u64, JS_MAX_SAFE_INTEGER);
+    let power = 10u64.pow((optimal_range as f64).log10().ceil() as u32);
 
-pub async fn process_module(
-    ident_str: RcStr,
-    full_hash: u64,
-    id_map: &mut HashMap<RcStr, Vc<ModuleId>>,
-    used_ids: &mut HashSet<u64>,
-) -> Result<()> {
-    if id_map.contains_key(&ident_str) {
-        return Ok(());
-    }
+    let mut module_id_map: HashMap<RcStr, Vc<ModuleId>> = HashMap::new();
+    let mut used_ids: HashSet<u64> = HashSet::new();
 
-    let mut trimmed_hash = full_hash % 10;
-    let mut power = 10;
-    while used_ids.contains(&trimmed_hash) {
-        if power >= u64::MAX / 10 {
-            // We don't want to take the next power as it would overflow
-            if used_ids.contains(&full_hash) {
-                return Err(anyhow::anyhow!("This is a... 64-bit hash collision?"));
-            }
-            trimmed_hash = full_hash;
-            break;
+    for (module_ident, full_hash) in merged_module_ids.iter() {
+        let mut trimmed_hash = full_hash % power;
+        while used_ids.contains(&trimmed_hash) {
+            trimmed_hash = hash_xxh3_hash64(trimmed_hash) % power;
         }
```

Comment:
hashing the *trimmed*_hash will lead to the same chain of conflicts. You could do the same with the full_hash that would work kind of, but we could also just hash the original input with a extra value, which is the safest option:

```suggestion
        let mut i = 0;
        while used_ids.contains(&trimmed_hash) {
            let mut hasher = Xxh3Hash64Hasher::new();
            module_ident.deterministic_hash(&mut hasher);
            i.deterministic_hash(&mut hasher);
            let full_hash = hasher.finish()
            trimmed_hash = full_hash % power;
            i += 1;
        }
```